### PR TITLE
[API] Allow routing to the cluster specified in the request_header for weighted cluster case.

### PR DESCRIPTION
**Pull request by tyxia**
Add `cluster header` to weighted cluster.  `name` and `cluster_header` will be oneof `cluster_specifiter` field.
This enable the ability of routing to the cluster specified in the request_header for weighted cluster (i.e. multiple upstreams) case.

Regarding the compatibility:
1) old config + new envoy --- It will just use `name` field as before.
2) new config + old envoy --- The new config with cluster_header will be rejected because validate.rules is enabled and it requires `name` with min_len.

Note: I have implemented the change and tested it with end-to-end
integration test/example locally. Those changes will be in a separate PR once this API change is approved.

Signed-off-by: Tianyu Xia <tyxia@google.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by yanavlasov — DISMISSED** [review]

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -320,9 +320,35 @@ message WeightedCluster {
...
-    // Name of the upstream cluster. The cluster must exist in the
-    // :ref:`cluster manager configuration <config_cluster_manager>`.
-    string name = 1 [(validate.rules).string = {min_len: 1}];
+    // Only one of *name* and *cluster_header* may be specified.
+    oneof cluster_specifier {
+      option (validate.required) = true;
+
+      // Name of the upstream cluster. The cluster must exist in the
+      // :ref:`cluster manager configuration <config_cluster_manager>`.
+      string name = 1 [
```
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:329]: You can't move this existing field into a oneof, because that's a breaking API change.   You need to leave this outside of the oneof for now.

The purpose of the `(udpa.annotations.field_migrate).oneof_promotion` annotation is to tell us to move these fields into the oneof if/when we ever release a new version of this API.  That's basically what you're supposed to use in this kind of case instead of adding a oneof.

So basically, I would just do the following here:

1. Add the `(udpa.annotations.field_migrate).oneof_promotion` annotation to the existing `name` field.
2. Add the new `cluster_header` field with the same annotation.
3. Add comments for both fields saying that only one of the two may be set.

Comment (tyxia) [reply api/envoy/config/route/v3/route_components.proto]: Thanks for the review! I have thought about backwards compatibility but looks like Envoy's rule is stricter even though it is safe in terms of protobuf wire compatibility

If we leave this outside of `oneof`,  I think we will need to remove validate.rules on `name` and `cluster_header` for now as well. Otherwise, Proto constraint validation will fail.

I updated the code as described above but I am new to Envoy API, if you have any other good ideas, please let me know.

File: api/envoy/config/route/v3/route_components.proto [hunk 2]
```diff
@@ -320,9 +320,29 @@ message WeightedCluster {
 
     reserved "per_filter_config";
 
+    // Only one of *name* and *cluster_header* may be specified.
+    // [#next-major-version: Need to add back the validation rule: (validate.rules).string = {min_len: 1}
```
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:324]: I think you need a `]` at the end of this line.

File: api/envoy/config/route/v3/route_components.proto [hunk 3]
```diff
@@ -320,9 +320,29 @@ message WeightedCluster {
...
     // Name of the upstream cluster. The cluster must exist in the
     // :ref:`cluster manager configuration <config_cluster_manager>`.
-    string name = 1 [(validate.rules).string = {min_len: 1}];
+    string name = 1 [(udpa.annotations.field_migrate).oneof_promotion = "cluster_specifier"];
+
+    // Only one of *name* and *cluster_header* may be specified.
+    // [#next-major-version: Need to add back the validation rule: (validate.rules).string = {min_len: 1 well_known_regex: HTTP_HEADER_NAME strict: false}
```
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:330]: Same here.

File: api/envoy/config/route/v3/route_components.proto (hunk 1 above)
Comment (markdroth) [reply api/envoy/config/route/v3/route_components.proto]: Moving a field into a oneof is fine for wire compatibility, but it will cause API changes in the generated code in golang.  Since we promise both wire compatibility and stability for build dependency for API clients other than Envoy (such as grpc-go), that would be a breaking change.

I suspect that removing the proto-gen-validate constraint is okay, as long as we add a check to the implementation to ensure that one of these two fields is set.  @envoyproxy/api-shepherds, can someone confirm that that's correct?

File: api/envoy/config/route/v3/route_components.proto (hunk 3 above)
Comment (tyxia) [reply api/envoy/config/route/v3/route_components.proto]: Done.

File: api/envoy/config/route/v3/route_components.proto (hunk 2 above)
Comment (tyxia) [reply api/envoy/config/route/v3/route_components.proto]: oops, missed `]` on this code comment. Thanks for catching it!
done.

File: api/envoy/config/route/v3/route_components.proto (hunk 1 above)
Comment (markdroth) [reply api/envoy/config/route/v3/route_components.proto]: @htuch says that Envoy will need to do explicit validation of these fields, so he'd prefer if the Envoy implementation was made part of this PR.  Otherwise, we will lose the PGV check without yet having the new explicit check.

**Review by yanavlasov — DISMISSED** [review]

File: api/envoy/config/route/v3/route_components.proto (hunk 1 above)
Comment (tyxia) [reply api/envoy/config/route/v3/route_components.proto]: Make sense.
In the spirit of "small cl", I have added the minimal implementation + validation logic and its unit tests (rather than adding full CL e.g. with weighted cluster pick logic, integration test, etc).

Please let me what do you think.

**Comment by tyxia** [comment]
Thanks for the review, @yanavlasov! I just fixed a format issue(i.e. change
free function name to start with lower case)

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -59,6 +59,27 @@ void mergeTransforms(Http::HeaderTransforms& dest, const Http::HeaderTransforms&
...
+    throw EnvoyException("Only one of name or cluster_header can be specified");
+  } else if (cluster.name().empty() && cluster.cluster_header().empty()) {
+    throw EnvoyException("At least one of name or cluster_header need to be specified");
+  } else {
+    if (!cluster.cluster_header().empty()) {
+      envoy::type::matcher::v3::RegexMatcher matcher;
+      const std::string non_strict_pattern = "^[^\\x{0000}\\x{000A}\\x{000D}]*$";
+      matcher.mutable_google_re2();
+      matcher.set_regex(non_strict_pattern);
+      if (!Regex::Utility::parseRegex(matcher)->match(cluster.cluster_header())) {
```
Comment (htuch) [inline source/common/router/config_impl.cc:74]: Basic idea here makes sense, but why can't you use a PGV annotation for `cluster_header` here? This is a new field, so we shouldn't have to do the HTTP header validation work manually, we can just use the annotation.

Comment (tyxia) [reply source/common/router/config_impl.cc]: Thanks for the review! 
We cannot do the full PGV annotation with `min_len: 1` here because it is not `oneof` field along with `name` yet. But we can just do PGV annotation for HTTP header validation part, which i agree it will be cleaner than manual RE2 validation in the implementation.

Done, I updated the corresponding unit test as well.  PTAL!

**Review by htuch — COMMENTED** [review]
Validation makes strategy to me 👍

**Comment by phlax** [comment]
@tyxia @yanavlasov @markdroth is this ready to land ?

**Comment by tyxia** [comment]
> @tyxia @yanavlasov @markdroth is this ready to land ?

Hi @phlax,
Yes, it is ready to land. I synced with @markdroth yesterday, it is waiting for the approval and merge from one of maintainers. @yanavlasov also approved this PR before (he is out of office this week).

File: api/envoy/config/route/v3/route_components.proto [hunk 5]
```diff
@@ -320,9 +320,31 @@ message WeightedCluster {
...
+    // .. attention::
+    //
+    //   Internally, Envoy always uses the HTTP/2 *:authority* header to represent the HTTP/1
+    //   *Host* header. Thus, if attempting to match on *Host*, match on *:authority* instead.
+    //
+    // .. note::
+    //
+    //   If the header appears multiple times only the first value is used.
+    string cluster_header = 12 [
+      (validate.rules).string = {well_known_regex: HTTP_HEADER_NAME strict: false},
```
[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto:345]: Yes, these were just the validators (PGV moral equivalent).
